=str add VirtualProcessor

rkuhn opened the pull request:
This is only for reviewing the idea, the name is of course wrong and it also needs tests (and be used of course).

rkuhn committed:
=str add VirtualProcessor

- create a fully RS compliant minimal identity processor (that is not an
  Actor)
- replace SubscriberSourceVirtualProcessor,
  PublisherSinkVirtualSubscriber and PublisherSinkVirtualPublisher with
  this
- add tests that transform Sink.publisher’s and Source.subscriber’s
  materialized value
- also remove the Keep.{left, right} optimization in order to make
  side-effecting mat value transforms execute even if their values are
  discarded

drewhk commented on akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala:
```diff
@@ -293,6 +293,51 @@ private[akka] object StreamLayout {
+  override def subscribe(s: Subscriber[_ >: T]): Unit =
+    if (subscriptionStatus.compareAndSet(null, s)) () // wait for onSubscribe
+    else {
+      val sub = subscriptionStatus.getAndSet(s).asInstanceOf[Subscription]
```
this will fail on the second subscribe call with a ClassCastException

rkuhn replied:
Right, this should be turned into a SpecViolation instead.

drewhk replied:
It is not a spec violation, just that this VirtualProcessor does not support fanout. This is btw one of the drawbacks of this approach, if you use Sink.publisher to access the features a possibly more powerful Publisher to the left, now it is reduced to single-subscription. I think it is fine, just for the record.

drewhk commented on akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala:
```diff
@@ -293,6 +293,51 @@ private[akka] object StreamLayout {
+    else {
+      val sub = subscriptionStatus.getAndSet(s).asInstanceOf[Subscription]
+      tryOnSubscribe(s, sub)
+      terminationStatus.getAndSet(Allowed) match {
```
this will be null if onSubscribe was first, since it does not set terminationStatus after the successful CAS on subscriptionStatus, and this will enter this branch since the CAS will fail here on subscriptionStatus. Maybe you wanted this to have the initial value of Allowed?

drewhk replied:
No, you just need to ignore the null, since setting it to Allowed in the getAndSet will force the onComplete to fail the CAS and call the onComplete itself

drewhk commented:
We can test it against the TCK by appending this processor to a real, actor backed identity processor:
 - attach to the left end
 - attach to the right end
 - attach to both ends

drewhk commented:
I have verified it, and it looks correct (I can present an informal "proof")

TODO:
 - test the mapMaterialized case where the Publisher/Subscriber is accessed and used in the mapper
 - add TCK tests with this appended to left, right, both ends of an IdentityProcessor
 - remove all the Virtual stuff and replace it with this one
 - remove and inline the attach() method in MaterializerSession
 - remove the createIdentityProcessor method in MaterializerSession

rkuhn commented:
okay, going forward with this now

rkuhn commented:
Implemented all the things and found a bonus bug as well (more correctly: an inconsistency in how `Keep.left` would completely drop the right computation instead of only discarding its value).

drewhk commented on akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala:
```diff
@@ -550,6 +620,7 @@ private[stream] abstract class MaterializerSession(val topLevel: StreamLayout.Mo
         case mv: MaterializedValueSource[_] ⇒
           val pub = new MaterializedValuePublisher
           materializedValuePublishers ::= pub
+          materializedValues.put(mv, ())
```
You should run the materialization benchmarks (or add a new one) to see how much impact does this have (given that the default is Keep.left almost always). If it is significantly slower ( > 5%) then I am against this.

rkuhn replied:
This is not optional: without it the tests for accessing the materialized value fail. I don’t expect much impact on real-world benchmarks since this only affects usage of `FlowGraph.Builder.materializedValue`.

drewhk replied:
Well, let's accept this as a feature then.

drewhk commented on akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala:
```diff
@@ -293,38 +293,108 @@ private[akka] object StreamLayout {
+      s.cancel()
+    }
+    @tailrec
+    override def request(n: Long): Unit = {
```
ewww. Why is this needed? There is something frightening about double accounting of requests.

rkuhn replied:
Without this the TCK fails: it rightfully signals reentrant signaling towards the Subscriber since requesting from onSubscribe may now immediately trigger the asynchronous invocation of onNext. Therefore requests during onSubscribe are delayed by this method.

drewhk replied:
Ok.

drewhk commented:
Please also change the materialization of the Identity module to use this instead of Transform(() => Map(identity))

rkuhn commented:
Concerning changing the materialization of Identity nodes: is the intention that these will be removed before fusing anyway? In that case I agree.

rkuhn committed:
=str fix sub-upstream cancellation in concatAll

- ActorProcessor terminated eagerly when ConcatAll had just taken up a
  new input stream but not yet received onSubscribe for it

- The ActorProcessor eagerly shuts itself down upon onError and that
  cannot be changed without completely reworking the Pump, so I opted
  for just tracking the outstanding substreamSubscribers that have not
  yet seen OnSubscribe and making them cancel properly when that arrives
  (possibly later).

rkuhn commented:
@drewhk I have replaced Identity by this VirtualProcessor and then by pure chance I saw the same failure as in https://jenkins.akka.io/job/pr-validator-per-commit-jenkins/2953/ locally: ConcatAllImpl suffers from stopping the ActorProcessor too eagerly (when onSubscribe has not yet been received from the substream). Please review the second commit carefully.

drewhk commented:
> Concerning changing the materialization of Identity nodes: is the intention that these will be removed before fusing anyway? In that case I agree.

These cannot be removed completely trivially in all cases, see

   incomingConneciton.flow.join(Flow[ByteString]) // echo server

I guess it is possible to detect this in the materialization phase, but fusing cannot help here.

rkuhn committed:
fix FlowStageSpec (and some boyscouting)

rkuhn commented:
Failure was remaining child actors in TlsSpec—is that one of the typical failure modes or a new one?

@drewhk Please review especially the last three commits since they fix problems that are unrelated to the VirtualProcessor itself, they were racy before and might have come up now due to timing changes.

drewhk commented:
yes, that is the typical failure.

drewhk commented:
mergeable?

rkuhn commented:
Yes, I’d say it is :-)

rkuhn commented:
Have you seen the improved fix in 5b420c5?

drewhk commented:
yes, please squash

rkuhn commented:
It is already squashed: the commits fix different things.
